src/NHSD.BuyingCatalogue.Identity.Api/Infrastructure/ValidationSummaryTagHelper.cs

```diff
@@ -0,0 +1,72 @@
+
+namespace NHSD.BuyingCatalogue.Identity.Api.Infrastructure
+{
+    [HtmlTargetElement(TagHelperConstants.Div, Attributes = TagHelperName)]
+    public sealed class ValidationSummaryTagHelper : TagHelper
+    {
+        public const string TagHelperName = "nhs-validation-summary";
+        public const string TitleName = "title";
+
+        [HtmlAttributeNotBound]
+        [ViewContext]
+        public ViewContext ViewContext { get; set; }
+
+        [HtmlAttributeName(TitleName)]
+        public string Title { get; set; }
+
+        public override void Process(TagHelperContext context, TagHelperOutput output)
+        {
+            output.ThrowIfNull();
+            if (ViewContext.ViewData.ModelState.IsValid)
+            {
+                output.Content.Clear();
+                return;
+            }
+
+            var errorSummary = new TagBuilder(TagHelperConstants.Div);
+            errorSummary.AddCssClass(TagHelperConstants.NhsValidationSummary);
+            errorSummary.Attributes[TagHelperConstants.Role] = TagHelperConstants.RoleAlert;
+            errorSummary.Attributes[TagHelperConstants.LabelledBy] = TagHelperConstants.ErrorSummaryTitle;
+
+            var header = new TagBuilder(TagHelperConstants.SubHeader);
+            header.AddCssClass(TagHelperConstants.NhsValidationSummaryTitle);
+            header.Attributes[TagHelperConstants.Id] = TagHelperConstants.ErrorSummaryTitle;
+            header.InnerHtml.Append(Title);
+            errorSummary.InnerHtml.AppendHtml(header);
+
+            var errorList = new TagBuilder(TagHelperConstants.UnorderedList);
+            errorList.AddCssClass(TagHelperConstants.NhsList);
+            errorList.AddCssClass(TagHelperConstants.NhsValidationSummaryList);
+            errorSummary.InnerHtml.AppendHtml(errorList);
+
+            var viewType = ViewContext.ViewData.Model.GetType();
+            viewType.ThrowIfNull();
+
+            var propertyNames = viewType.GetProperties().Select(x => x.Name).ToList();
+            var orderedStates = ViewContext.ViewData.ModelState
+                .OrderBy(d => propertyNames.IndexOf(d.Key))
+                .ToList();
+
+            foreach (var model in orderedStates)
+            {
+                foreach (var error in model.Value.Errors)
+                {
+                    var listItem = new TagBuilder(TagHelperConstants.ListItem);
+                    var errorElement = new TagBuilder(TagHelperConstants.Anchor);
+                    errorElement.Attributes.Add(TagHelperConstants.Link, $"#{model.Key}");
+                    errorElement.InnerHtml.Append(error.ErrorMessage);
+                    listItem.InnerHtml.AppendHtml(errorElement);
+                    errorList.InnerHtml.AppendHtml(listItem);
+                }
+            }
+
+            output.TagName = TagHelperConstants.Div;
+            output.TagMode = TagMode.StartTagAndEndTag;
+            output.Content.AppendHtml(errorSummary);
+        }
```

Comment:
As above, might it be worth moving the build of some of the inner elements out into separate methods, e.g. header, list etc? (Purely for readability.)